build/org.eclipse.cdt.build.gcc.core/src/org/eclipse/cdt/build/gcc/core/GCCToolChain.java

```diff
@@ -640,42 +645,102 @@ public String[] getCompileCommands(ILanguage language) {
 		}
 	}
 
+	/**
+	 * Construct regular expression to find any file extension for C or C++. Returns
+	 * expression shaped in form of "((cpp)|(c++)|(c))".
+	 * Copied from {@link AbstractLanguageSettingsOutputScanner}.
+	 *
+	 * @return regular expression for searching C/C++ file extensions.
+	 */
+	private String getPatternFileExtensions() {
+		IContentTypeManager manager = Platform.getContentTypeManager();
+
+		Set<String> fileExts = new HashSet<>();
+
+		IContentType contentTypeCpp = manager.getContentType(CCorePlugin.CONTENT_TYPE_CXXSOURCE);
+		fileExts.addAll(Arrays.asList(contentTypeCpp.getFileSpecs(IContentType.FILE_EXTENSION_SPEC)));
+
+		IContentType contentTypeC = manager.getContentType(CCorePlugin.CONTENT_TYPE_CSOURCE);
+		fileExts.addAll(Arrays.asList(contentTypeC.getFileSpecs(IContentType.FILE_EXTENSION_SPEC)));
+
+		String pattern = expressionLogicalOr(fileExts);
+
+		return pattern;
+	}
+
+	/**
+	 * Helper method to construct logical "or" to be used inside regular
+	 * expressions.
+	 * Copied from {@link AbstractLanguageSettingsOutputScanner}.
+	 */
+	@SuppressWarnings("nls")
+	private static String expressionLogicalOr(Set<String> fileExts) {
+		String pattern = "(";
+		for (String ext : fileExts) {
+			if (pattern.length() != 1)
+				pattern += "|";
+			pattern += "(" + Pattern.quote(ext) + ")";
+			ext = ext.toUpperCase();
+			if (!fileExts.contains(ext)) {
+				pattern += "|(" + Pattern.quote(ext) + ")";
+			}
+		}
+		pattern += ")";
+		return pattern;
+	}
+
+	private static String getFileExtension(String fileName) {
+		String ext = ""; //$NON-NLS-1$
+
+		int i = fileName.lastIndexOf('.');
+		if (i > 0) {
+			ext = fileName.substring(i + 1);
+		}
+		return ext;
+	}
+
 	@Override
 	public IResource[] getResourcesFromCommand(List<String> cmd, URI buildDirectoryURI) {
 		// Start at the back looking for arguments
 		List<IResource> resources = new ArrayList<>();
 		IWorkspaceRoot root = ResourcesPlugin.getWorkspace().getRoot();
+		String extensionPattern = getPatternFileExtensions();
+
 		for (int i = cmd.size() - 1; i >= 0; --i) {
 			String arg = cmd.get(i);
 			if (arg.startsWith("-")) { //$NON-NLS-1$
-				// ran into an option, we're done.
-				break;
+				// ran into an option, skip.
+				continue;
 			}
 			if (i > 1 && cmd.get(i - 1).equals("-o")) { //$NON-NLS-1$
 				// this is an output file
 				--i;
 				continue;
 			}
-			try {
-				Path srcPath = Paths.get(arg);
-				URI uri;
-				if (srcPath.isAbsolute()) {
-					uri = srcPath.toUri();
-				} else {
-					String mingwPath = fixMingwPath(arg);
-					if (mingwPath != arg) {
-						uri = Paths.get(mingwPath).toUri();
+
+			String ext = getFileExtension(arg);
+			if (Pattern.matches(extensionPattern, ext)) {
```

Comment:
I don't think a regular expression is the suitable data structure here. How about something like this:

```java

	private Set<String> getPatternFileExtensions() {
		IContentTypeManager manager = Platform.getContentTypeManager();

		Set<String> fileExts = new TreeSet<String>(String.CASE_INSENSITIVE_ORDER);

		IContentType contentTypeCpp = manager.getContentType(CCorePlugin.CONTENT_TYPE_CXXSOURCE);
		fileExts.addAll(Arrays.asList(contentTypeCpp.getFileSpecs(IContentType.FILE_EXTENSION_SPEC)));

		IContentType contentTypeC = manager.getContentType(CCorePlugin.CONTENT_TYPE_CSOURCE);
		fileExts.addAll(Arrays.asList(contentTypeC.getFileSpecs(IContentType.FILE_EXTENSION_SPEC)));

		return fileExts;
	}

```

and then here do:

```java
    if (getPatternFileExtensions().contains(ext)) {
```

(although it may be best to cache the getPatternFileExtensions result somewhere as the number of arguments and number of command lines to be processed could be quite large)

Reply:
Hi Jonah,
Thanks for your comment. Regular expression is indeed overkill here. I will update the change.
kind regards,
Erwin